sympy/calculus/util.py

```diff
@@ -1348,17 +1353,61 @@ def __pow__(self, other):
-                    if den % 2 == 0:
-                        if S.Zero in self:
-                            if self.min.is_extended_negative:
-                                return AccumBounds(0, real_root(self.max, den))
-                    return AccumBounds(real_root(self.min, den),
-                                       real_root(self.max, den))
-                if den!=1:
-                    num_pow = self**num
-                    return num_pow**(1 / den)
-            return AccumBounds(-oo, oo)
+                if num is S.One:
+                    return AccumBounds(*[i**(1/den) for i in self.args])
+
+                elif den is not S.One:  # e.g. if other is not Float
+                    return (self**num)**(1/den)  # ok for non-negative base
+
+            if isinstance(other, AccumBounds):
+                if (
+                        self.min.is_zero is not None and
+                        other.min.is_extended_negative is not None):
+                    # negative numbers require only integer exponents to produce
+                    # real bounds and AccumBounds includes non-integers so we
+                    # must deal only with nonnegative bases in self.
+                    if other.min.is_extended_negative:
+                        if self.max.is_extended_negative:
+                            return S.NaN  # finite complex values
+                        elif 0 in self:
+                            return S.ComplexInfinity
+                    p = set()
+                    if self.min.is_zero and other.max.is_extended_positive:
+                        p.add(0)
+                    if 0 in other:
+                        p.add(1)  # x**0 = 1
+                    if 1 in self:
+                        p.add(1)  # 1**x = 1
+                    b = set(self.args) - {0}  # handled above
+                    e = set(other.args)
+                    # compute candidate powers
+                    p |= {i**j for i in b for j in e}
+                    return self.func(min(p), max(p))
+
+            raise NotImplementedError  # to leave unevaluated instead of return AccumBounds(-oo, oo)
+
+        return NotImplemented
+
+    @_sympifyit('other', NotImplemented)
+    def __rpow__(self, other):
+        if other.is_real and other.is_nonnegative:
+            if other is S.One:
+                return S.One
+            p = set()
+            if other.is_extended_positive:
+                p.add(other**self.min)
+                p.add(other**self.max)
+            elif other.is_zero:
+                if self.min.is_zero:
+                    p.add(0)
+                    if self.max.is_extended_positive:
+                        p.add(1)
+                elif self.min.is_extended_positive:
+                    p.add(1)
+                elif self.max.is_extended_nonpositive:
+                    return S.ComplexInfinity
+        if p:
+            return self.func(min(p), max(p))
 
         return NotImplemented
```

Comment:
The TypeError for `z = 0**B(a, a + S.Half)` comes from this line: nothing was added to `p`.